### DISCUSSION TITLE
expo push notifications

### DISCUSSION THREAD

**Discussion by EdgarsLv** [supabase/supabase#8745] [Questions] [answered]
``Need help for expo push notifications.

Basically i need this function to be triggered on message insert. 
![image](https://user-images.githubusercontent.com/56837461/187768500-a1ce2803-48ad-484c-b6a0-b988a66cc5ad.png)

I enabled http extension and was thinking about somethig like this:

![image](https://user-images.githubusercontent.com/56837461/187769083-45ac2afb-b841-4482-8a92-a4709d8d8b15.png)

but of course this code is wrong(took it from example and modified). have read available documentation(also github official http extension page but i'm too noobish to understand it) and still don't  even understand what that select STATUS and CONTENT even does.

users.push_token and users.name should come from users table matching messages.receiver_id - also not sure how to get them...

I will be grateful for any help.

**Comment by GaryAustin1** [comment, marked as answer]
A few suggestions to get you closer.
I believe for http_post the third parameter is content type.  If you need headers you probably have to use the http() method.
You will need to set search_path =  extensions,public.
To get your message columns you would use new.col.
To get user table data you will need to select from it with a user id from messages if you have one.  Otherwise you probably have auth.uid() for the user inserting the message.

**Reply by GaryAustin1** [reply]
In response to below, There is probably an error in that function.  
Two ways to debug:
One is to look at database postgres log in the dashboard for errors. 
Second is to run the select part in the SQL window to test it.

One thing I see is you don't have a return at end which a trigger function needs.   `RETURN new;` (or `RETURN null;` if an after trigger) after the SELECT statement is needed. 

Also it is better to add comments on to your original post, or answers than to create existing answers when you are clarifying or asking for more help.

**Reply by EdgarsLv** [reply]
don't think it has anything to do with function not working. btw can't invoke trigger function with select.
created this simple trigger:
![image](https://user-images.githubusercontent.com/56837461/188273556-08b00bc8-912d-4499-9a62-2d35a0ab008e.png)

don't understand what to look for in Postgres log, but in API log i'm getting 400 or 404 status error: 
![image](https://user-images.githubusercontent.com/56837461/188273787-a985f15d-8143-4a5b-bbcc-b58ac26bbb2c.png)

ps.trigger is on messages table

**Reply by GaryAustin1** [reply]
At a minimum search path should have extensions not extension.
There should be error messages in database tab Postgres logs.

**Reply by EdgarsLv** [reply]
changing to extanesions didn't help.

getting this in network tab and also same in Postgress log
![image](https://user-images.githubusercontent.com/56837461/188287767-1b1d189c-03b5-4f0f-8ecb-52538718ec54.png)

edit. changing SELECT to PERFORM fixed error. Now getting back to notification function

**Reply by EdgarsLv** [reply]
ok got trigger to work and notifications are sent. 
now just to figure out how dynamically pass correct data...

**Reply by GaryAustin1** [reply]
The data will come in the new object.  So new.id would be the inserted id column.
You then have to format your http output as either url parameters or a post body.  You could use string substitution or concatenation to build things up, or possibly build a json structure if going into a post body.

**Reply by EdgarsLv** [reply]
this works, but i can't pass any dynamic data here... can't interpolate, concantenate or whatever...
![image](https://user-images.githubusercontent.com/56837461/188291272-6b90b638-81c4-4bb3-a6f8-69d1c4f2af54.png)

found this way and at least can pass dynamic data:
![image](https://user-images.githubusercontent.com/56837461/188291334-40556a92-c1fc-4cb2-9bfc-1b867693c4b0.png)
but getting this error:
![image](https://user-images.githubusercontent.com/56837461/188291354-5c14f41a-647e-4fb4-8aac-94a6bca5420e.png)
while documentation says:
![image](https://user-images.githubusercontent.com/56837461/188291366-6c49d0dc-15ca-4667-8551-b12ba3365d6f.png)

**Reply by GaryAustin1** [reply]
So I have a function that uses the storage api and build up my string using concatenation into a variable and then use the variable....
```
declare
  myurl varchar := 'https://zkixxxxxxxxxxxblxrq.supabase.co';
  service_role_key varchar := 'longjwtstring';
  url varchar := myurl || '/storage/v1/object/' || bucket || '/' || pathname;
begin
return (
select status FROM http((
          'DELETE',
            url,
            .............
```
bucket and pathname are passed to the function, but could have been new.bucket for example.

But I'm at the point I'd have to google to figure out how to format a json body.
You might want to ask a new question as not many will look after this many replies.

**Reply by EdgarsLv** [reply]
got it to work like this: 
![image](https://user-images.githubusercontent.com/56837461/188309788-b9371377-254c-4ffc-a6c4-fc3430f10b52.png)

Now just need few selects as token comes from users table which i can't get with new keyword and here i used hard-coded in variable.

And ifs to check if message receiver has token to invoke trigger as only app users will have token but not web users.
Hope that is even possible?

**Reply by EdgarsLv** [reply]
Finished function and everything works well. Just not sure how good or bad this function is...

Just in case this can help someone else:
![image](https://user-images.githubusercontent.com/56837461/188335472-733f58c7-1b25-4bfb-8bde-7a9dc2cd0bb4.png)

And thank you for help.

**Reply by GaryAustin1** [reply]
Looks reasonable.  At some point you may want to actually get the status back from the http call and at a minimum log it if it is an error.
You can generate log messages with raise log 'push send got an error:  %', error;   //error is from the http return obj.

**Reply by javierguzman** [reply]
Hello @EdgarsLv  I just wanted to say thank you for leaving your code here for people in the future like me who does not know how to solve this!

**Comment by EdgarsLv** [comment]
So based on this: 
![image](https://user-images.githubusercontent.com/56837461/188208485-e26d9ff0-2c01-4d4a-b565-eaf99fd54a10.png)

created this function with hard coded value and token, not sure if it's correct, but was able to create it. Also created trigger on messages table to test if works. 

Problem, that now i can't  insert anything in messages table. Why is that?
![image](https://user-images.githubusercontent.com/56837461/188208411-15065f03-1416-49fc-b451-0258d71ee6e1.png)

**Comment by sotirelisc** [comment]
I found it easier & quicker to implement by using Supabase Edge Functions and Database Webhooks instead of Database Triggers.
This is the process that I follow:

1. Get device Expo push notifications token (check Expo documentation for that) & store it on Supabase.
2. Create a Supabase Edge Function & Database Webhook to call this function whenever a record is inserted (or other action). For this I followed the OneSignal integration example by Supabase (https://supabase.com/docs/guides/integrations/onesignal) but instead of using OneSignal, I just execute a POST request to Expo with all the necessary data to send a push notification.

**Reply by GlendaleAcosta** [reply]
Had the same idea! Seems to be working well so far.

**Reply by albertpco** [reply]
I got to hear but am getting 502 and 400 errors. Might be misunderstanding the structure but any help would be amazing. Currently have a trigger set up, edge function, and it’s tied to Postgres as I need more info than just from the table with a change to send a better notification.
